app/src/gui/components/map/tile-source.ts

```diff
@@ -0,0 +1,588 @@
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND either express or implied.
+ * See, the License, for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Description:
+ *   Implements a tile source for Open Street Map tiles that can cache
+ * tiles locally.
+ */
+
+import {Extent} from 'ol/extent';
+import {FeatureLike} from 'ol/Feature';
+import MVT from 'ol/format/MVT';
+import TileLayer from 'ol/layer/Tile';
+import VectorTileLayer from 'ol/layer/VectorTile';
+import {LoaderOptions} from 'ol/source/DataTile';
+import ImageTileSource from 'ol/source/ImageTile';
+import OSM, {ATTRIBUTION} from 'ol/source/OSM';
+import VectorTileSource from 'ol/source/VectorTile';
+import {TileCoord} from 'ol/tilecoord';
+import VectorTile from 'ol/VectorTile';
+import {MAP_SOURCE, MAP_SOURCE_KEY, MAP_STYLE} from '../../../buildconfig';
+import {applyStyle} from 'ol-mapbox-style';
+import {getMapStylesheet} from './styles';
+import Tile from 'ol/Tile';
+
+// Table of map tile sources for raster and vector tiles
+// based on configuration settings we select which of these to use
+//
+const TILE_URL_MAP: {[key: string]: {[key: string]: string}} = {
+  osm: {
+    raster: 'https://tile.openstreetmap.org/{z}/{x}/{y}.png',
+    vector: 'https://tile.openstreetmap.org/data/{z}/{x}/{y}.pbf',
+  },
+  maptiler: {
+    raster:
+      'https://api.maptiler.com/maps/outdoor-v2/{z}/{x}/{y}.png?key={key}',
+    vector: 'https://api.maptiler.com/tiles/v3/{z}/{x}/{y}.pbf?key={key}',
+  },
+};
+
+interface StoredTile {
+  url: string;
+  data: Blob;
+  sets: string[];
+}
+
+export interface StoredTileSet {
+  setName: string;
+  extent: number[];
+  minZoom: number;
+  maxZoom: number;
+  size: number;
+  expectedTileCount: number;
+  created: Date;
+  tileKeys: IDBValidKey[];
+}
+
+type KeyType = string | string[] | null | undefined;
+class IDB<Type> {
+  keyPath: KeyType;
+  db: any;
+  dbName: string;
+
+  constructor(db: any, dbName: string, keyPath: KeyType) {
+    this.keyPath = keyPath;
+    this.db = db;
+    this.dbName = dbName;
+  }
+
+  // Create the object store for this database, called in
+  // onupgradeneeded for the database
+  createObjectStore() {
+    if (!this.db.objectStoreNames.contains(this.dbName)) {
+      this.db.createObjectStore(this.dbName, {
+        keyPath: this.keyPath,
+      });
+    }
+  }
+
+  async get(query: IDBKeyRange | IDBValidKey): Promise<Type | undefined> {
+    return new Promise<Type | undefined>((resolve, reject) => {
+      if (!this.db) {
+        return;
+      }
+      const transaction = this.db.transaction(this.dbName, 'readonly');
+      const store = transaction.objectStore(this.dbName);
+      const request = store.get(query);
+      request.onsuccess = () => {
+        resolve(request.result);
+      };
+      request.onerror = () => reject(request.error);
+    });
+  }
```

Comment:
Generally this class needs to be well documented given some non-obvious indexedDB stuff going on